glean/metrics.yaml

```diff
@@ -1637,3 +1637,179 @@ sample:
       language_code:
         description: The language code chosen.
         type: string
+
+  speed_test_completed:
```

Comment:
From the PRD I understand this is coming from the "# of times a speed test is performed" request. Correct?

I would suggest adding a metric for `speed_test_started` as well so we can have a full picture of speed test usage, even if it doesn't complete.

Reply:
I had that originally, but pulled it based on [Basti's feedback above](https://github.com/mozilla-mobile/mozilla-vpn-client/pull/5702#discussion_r1084343827). 

I'll ask Santiago on the PRD what his hope is for this one.

Reply:
This circles back to, what exactly do we want to answer with this metric? 

Do we want to understand the usage of the speed test? In that case we would need the "started" metric to get a full picture. Things we may want to know: How many users attempt a speed test and leave before it completes? How long does it take for a speed test to complete? How long before a user gives up? What do users do after the speed test? What events precede a speed test attempt? This metric alone does not allow us to answer all of this. 

With the current approach, we can only say "when a speed test is completed, the result is X". I don't see too much value in that. It looks more like a performance metric, and if we were interested in performance, interaction metrics are likely not the best approach either.

Reply:
I've added logging for both errors and user cancellations. Thanks for the nudge, I agree it's much stronger with that.

It took a bit, because we were initially logging errors with every user cancellation OR missing some errors, if we logged errors in a different spot. 

I believe the "logging errors with each cancellation" was happening because we were incorrectly considering a user-cancelled DNS lookup an error. (A user would cancel the speed check, we'd intentionally cancel the upload and download tasks, but a DNS error of user-cancellation would still trigger the code to set an error state.) I believe I've cleared this up in `benchmarktasktransfer.cpp`.

I believe I'm currently only logging one end state for the end of any speed test, but since there are many ways they could end, I didn't see a single place in the code to handle all cases, unless I added a new Cancelled state to ConnectionBenchmark, which seemed heavy handed and would bring risks.